### PR TITLE
Victory Point charts with prediction

### LINKED ISSUE #5: display projected victory points?
would it make sense to show the projected victory points?

meaning
1. current points
2. `- bid`
3. `+ final points if all current resources were converted to c`

### PR DESCRIPTION
**Pull request by zeitlinger**
add victory points chart

Fixes #5 

image

Screenshot 2021-02-28 15:59:32

### CODE REVIEW

File: viewer/src/components/Charts.vue [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+@Component({
+  components: { PlayerCircle },
+})
+export default class Charts extends Vue {
+
+  data() {
+    return {
+      selected: PlayerEnum.All,
+    };
+  }
```
Comment (coyotte508) [inline viewer/src/components/Charts.vue:45]:
```suggestion
  selected = PlayerEnum.All
```

If you'd really wanted to use data(), it would have been better in the `@Component` part, but there's no reason to:

```ts
@Component({
  data() {
    // values returned are then usable directly with `this`
    return {...}
   }
}
```

With the Typescript class syntax for vue, you can just declare member variables with an initial value instead.

File: viewer/src/components/Charts.vue [hunk 2]
```diff
@@ -0,0 +1,98 @@
+<template>
+  <div class="gaia-viewer-modal">
+    <div id="player-selection">
+      <span
+        v-for="index in players"
+        :key="index"
+        @click="
+          $data.selected = index;
+          selectPlayer(index);
```
Comment (coyotte508) [inline viewer/src/components/Charts.vue:9]: No need to use `$data`. And better to do it in `selectPlayer`, with `this.selected = index`

File: viewer/package.json [hunk 3]
```diff
@@ -28,9 +29,12 @@
   ],
   "dependencies": {
     "@gaia-project/engine": "workspace:../engine",
+    "@types/chart.js": "^2.9.31",
     "bootstrap": "^4.4.1",
     "bootstrap-vue": "^2.12.0",
+    "chart.js": "^2.9.4",
     "lodash": "^4.17.15",
+    "moment": "^2.29.1",
```
Comment (coyotte508) [inline viewer/package.json:37]: Is this used anywhere?

File: engine/src/player.ts [hunk 4]
```diff
@@ -595,11 +596,11 @@ export default class Player extends EventEmitter {
     }
   }
 
-  receivePassIncome() {
+  receivePassIncome(consume: (rewards: Reward[], source: EventSource) => void) {
     // this is for pass tile income (e.g. round boosters, adv tiles)
     for (const event of this.events[Operator.Pass]) {
       const times = this.eventConditionCount(event.condition);
-      this.gainRewards(
+      consume(
```
Comment (coyotte508) [inline engine/src/player.ts:603]: Why this change? Is `receivePassIncome` called somewhere else?

**Review by coyotte508 — CHANGES_REQUESTED** [review]
I'm really impressed by this PR:

image

And then clicking a player shows the detailed breakdown. It's really good.

Since this PR is a bit complex, please make new commit(s) for the review fixes, I will squash the commit when merging.

File: viewer/package.json [hunk 5]
```diff
@@ -10,6 +10,7 @@
   "scripts": {
     "serve": "vue-cli-service serve --port 8080",
     "build": "vue-cli-service build",
+    "compile": "tsc",
```
[the remark this answers is not in the document]
Comment (zeitlinger) [reply viewer/package.json:13]: not really useful - I'll remove it.

File: viewer/src/components/Charts.vue (hunk 1 above)
Comment (zeitlinger) [reply viewer/src/components/Charts.vue:45]: I thought this would somehow not work - but it was probably just the auto-complete that didn't work..

File: viewer/package.json (hunk 3 above)
Comment (zeitlinger) [reply viewer/package.json:37]: chart.js threw a runtime error - the new version of chart.js (which is still beta) doesn't require it anymore.

File: engine/src/player.ts (hunk 4 above)
Comment (zeitlinger) [reply engine/src/player.ts:603]: yes, in [vp estimation](https://github.com/boardgamers/gaia-project/pull/18/files#diff-d81e83bc496b6ae3460f7a869ee73a1e9011102be8df59ef70f5d81b671cd028R84)

File: viewer/src/components/Game.vue [hunk 6]
```diff
@@ -132,7 +132,7 @@ import { parseCommands } from "../logic/recent";
     ScoringBoard,
     SpaceMap,
     TurnOrder,
-    Charts,
+    Charts: () => import('./Charts.vue'),
```
Comment (coyotte508) [inline viewer/src/components/Game.vue:135]: You also need to remove the `import Charts from "./Charts.vue";` at the top of the scripts for the dynamic load to be meaningul

File: engine/src/player.ts (hunk 4 above)
Comment (coyotte508) [reply engine/src/player.ts:603]: I understand, but I'd like to avoid making the engine code more complicated for things specific to the viewer. Here it's a bit hard to understand the reasoning without at least a comment, and the style changes from the rest of the engine code. 

Luckily, there's an alternative which does not add too much complexity. When simulating VP, instead of cloning the player data, you can clone the player:

```
const json = JSON.parse(JSON.stringify(player));
delete json.federationCache; // otherwise we need to pass the map when loading
const clone = Player.fromData(json, null, 0);
```

That way you can just do `clone.receivePassIncome()` (and `clone.data.finalHandling()`, ...)

Comment (zeitlinger) [reply engine/src/player.ts]: tried it - but got a strange error - it's in the last commit. can you help?

Comment (zeitlinger) [reply engine/src/player.ts]: fixed it - I think it was due to the lazy loading of chart.

File: viewer/package.json (hunk 3 above)
Comment (coyotte508) [reply viewer/package.json:35]: We should use the beta version then

https://bundlephobia.com/result?p=moment@2.29.1
https://bundlephobia.com/result?p=chart.js@3.0.0-beta.12

Not only the chart.js beta is much smaller than the 2.9 version, it also does not require moment which is bulky (and outdated, now there are `date-fns` and `luxon`...).

It's also treeshakable, meaning we can only import parts we need and the rest of the code will not be included in the bundle.

https://www.chartjs.org/docs/master/getting-started/v3-migration

I can help in the migration / do the migration if you want

Comment (zeitlinger) [reply viewer/package.json:35]: would be great if you can help :smile:  - I already tried the beta, but couldn't get it to work.

Comment (coyotte508) [reply viewer/package.json:35]: I did most of the changes there: https://github.com/boardgamers/gaia-project/commit/bb40da5a5a5de6b5d631828b9305a0028e799d9f

Some things to fixes with tooltips probably

File: viewer/src/components/BoardAction.vue [hunk 7]
```diff
@@ -79,7 +79,7 @@ export default class BoardAction extends Vue {
 
   get recent(): boolean {
     const moves = this.$store.getters["gaiaViewer/recentCommands"];
-    return moves.some((c) => c.command == Command.Action && BoardAction[c.args[0]] === this.action);
+    return moves.some((c) => c.command == Command.Action && c.args[0] == this.action);
```
Comment (coyotte508) [inline viewer/src/components/BoardAction.vue:82]: Should be `===` for scalar types like strings.
